### PR TITLE
feat(run): allow-all permission prompts

### LINKED ISSUE #16550: Suggestion: `yes to all` option for specific permission prompts like 'env'
Some Deno modules need permissions to many different things, e.g. lots of different environment variables. When I'm working with a module that I don't know yet and using the permission prompts I sometimes have to confirm a lot of different prompts. When this happens usually I cancel `deno run` and retry with an option like `--allow-env`, but this is a bit of a cumbersome developer experience because I never quite know in advance how many prompts there are going to be (should I keep going with the prompts or re-run with the --allow-* flag?). For example this is how it can sometimes look on the terminal:

```
Download ⠴ https://deno.land/x/aws_api/services/sts/mod.ts
Warning Implicitly using latest version (v0.7.0) for https://deno.land/x/aws_api/services/sts/mod.ts
✅ Granted env access to "AWS_REGION".
✅ Granted env access to "AWS_ACCESS_KEY_ID".
✅ Granted env access to "AWS_SECRET_ACCESS_KEY".
✅ Granted env access to "AWS_SESSION_TOKEN".
✅ Granted env access to "AMAZON_ACCESS_KEY_ID".
✅ Granted env access to "AMAZON_SECRET_ACCESS_KEY".
✅ Granted env access to "AMAZON_SESSION_TOKEN".
✅ Granted env access to "AWS_SHARED_CREDENTIALS_FILE".
✅ Granted env access to "HOME".
✅ Granted env access to "AWS_PROFILE".
✅ Granted read access to "undefined/.aws/credentials".
✅ Granted env access to "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI".
✅ Granted env access to "AWS_CONTAINER_CREDENTIALS_FULL_URI".
✅ Granted env access to "AWS_CONTAINER_SERVICE_ENDPOINT".
⚠️  ┌ Deno requests env access to "AWS_CONTAINER_AUTHORIZATION_TOKEN".
   ├ Run again with --allow-env to bypass this prompt.
   └ Allow? [y/n] (y = yes, allow; n = no, deny) >
```

It would be nice to have a 'yes to all' option during the prompts so that all prompts for the same permission like 'env' or 'read' are accepted. Something like:

```
⚠️  ┌ Deno requests env access to "AWS_REGION".
   ├ Run again with --allow-env to bypass this prompt.
   └ Allow? [y/n/A] (y = yes, allow; n = no, deny, A = yes to all) > <-- accept all 'env' prompts from this point
```

The permissions being granted should then still be shown so that users can figure out the (minimal) permissions for a production script or Docker image.

### PR DESCRIPTION
**Pull request by iuioiua**
Previously, the user would be prompted for every new value within a given unary permission. This would result in the following output:
```
Download ⠴ https://deno.land/x/aws_api/services/sts/mod.ts
Warning Implicitly using latest version (v0.7.0) for https://deno.land/x/aws_api/services/sts/mod.ts
✅ Granted env access to "AWS_REGION". // Enter `y`
✅ Granted env access to "AWS_ACCESS_KEY_ID". // Enter `y`
✅ Granted env access to "AWS_SECRET_ACCESS_KEY". // Enter `y`
...
```

This change introduces a new option for unary runtime permission prompts: `A` to allow all values of the permission in question. This will result in the following output:
```
Download ⠴ https://deno.land/x/aws_api/services/sts/mod.ts
Warning Implicitly using latest version (v0.7.0) for https://deno.land/x/aws_api/services/sts/mod.ts
✅ Granted all env access. // Enter `A`
```

Closes #16550.

### CODE REVIEW

File: runtime/permissions/prompter.rs
Comment (bartlomieju) [inline runtime/permissions/prompter.rs:190]: Feel like this should be a boolean passed to `permission_prompt` instead of doing this check here

**Review by bartlomieju — COMMENTED** [review]
> Tests (any guidance involving user input such as Y and N?)

It would probably be easiest to test this in `repl_tests`

File: runtime/permissions/prompter.rs
Comment (iuioiua) [reply runtime/permissions/prompter.rs:190]: Do you mean where `is_unary_permission()` is defined or used? I thought where I had it defined and used made sense, as it's used to adjust the prompt message (see line 209) and handle invalid prompt responses (see lines 253 and 262).

Comment (bartlomieju) [reply runtime/permissions/prompter.rs:190]: I mean that `permissions_prompt` should have following signature:
```
pub fn permission_prompt(
  message: &str,
  flag: &str,
  api_name: Option<&str>,
  is_unary: bool
) -> PromptResponse {
```
Matching against string names seems a bit brittle to me. But maybe I'm overreacting here.

Comment (iuioiua) [reply runtime/permissions/prompter.rs:190]: Oh, I see. Yeah, that's definitely a better way to do it.

**Comment by bartlomieju** [comment]
@iuioiua this PR looks good and I would love to land it for 1.31. Could you rebase the PR?

**Comment by iuioiua** [comment]
Will do! However, I'd like to have a test for each permission before landing. Do you agree?

**Comment by bartlomieju** [comment]
Yes, that sounds reasonable

**Comment by iuioiua** [comment]
Hm... Why are other tests such as `_090_run_permissions_request` passing in CI? They shouldn't be...

File: runtime/permissions/prompter.rs
Comment (bartlomieju) [inline runtime/permissions/prompter.rs:209]: Some feedback here: `y/Y` and `n/N` are very confusing. I suggest to use other letters for "allow all"/"deny all" - how about "a" and "d" respectively? I think that might also fix problem with the tests you're having

Comment (iuioiua) [reply runtime/permissions/prompter.rs:209]: Good idea. But I don't think it'll fix the tests. They should be failing right but aren't irrespective of this change.

Comment (iuioiua) [reply runtime/permissions/prompter.rs]: Done.

**Review by aapoalas — CHANGES_REQUESTED** [review]
Aside from requiring a rebase, I tested the branch and indeed both upper and lower case Y and N work. This is I believe by design and also how it works with current release Deno CLI.

However, I noticed that the `A` / `a` option while being recognised in both lower and upper case doesn't necessarily work as one might expect.

With a file like this (this is from `tools/format.js`):
```ts
const cmd = new Deno.Command(execPath, {
  args: [subcommand, "--config=" + configFile],
  cwd: ROOT_PATH,
  stdout: "inherit",
  stderr: "inherit",
});

const { code } = await cmd.output();
```
A prompt will be given at the `cmd.output()` line to run the command. When given with `a` or `A` the command will then fail:

> error: Uncaught PermissionDenied: Requires run access to ...

Not sure if what that's about.

**Comment by iuioiua** [comment]
> Aside from requiring a rebase, I tested the branch and indeed both upper and lower case Y and N work. This is I believe by design and also how it works with current release Deno CLI.

Rebased. Did you mean `Y`/`N` allow/deny all permissions? If so, `Y` and `N` behaviour now haven't been affected by this PR.

Also, I tried the `a` behaviour, and it seems to work within expectations. Using `a` yields the following:
```
asher@Ashers-MacBook-Air deno % cargo run run /Users/asher/GitHub/deno/cli/tests/testdata/run/permissions_prompt_allow_deny_all.ts
    Finished dev [unoptimized + debuginfo] target(s) in 0.26s
     Running `target/debug/deno run /Users/asher/GitHub/deno/cli/tests/testdata/run/permissions_prompt_allow_deny_all.ts`
✅ Granted all run access.
✅ Granted all read access.
✅ Granted all write access.
✅ Granted all net access.
✅ Granted all env access.
✅ Granted all sys access.
✅ Granted all ffi access.
```

Using `d` yields the following:
```
asher@Ashers-MacBook-Air deno % cargo run run /Users/asher/GitHub/deno/cli/tests/testdata/run/permissions_prompt_allow_deny_all.ts
    Finished dev [unoptimized + debuginfo] target(s) in 0.24s
     Running `target/debug/deno run /Users/asher/GitHub/deno/cli/tests/testdata/run/permissions_prompt_allow_deny_all.ts`
❌ Denied all run access.
❌ Denied all run access.
❌ Denied all read access.
❌ Denied all read access.
❌ Denied all write access.
❌ Denied all write access.
❌ Denied all net access.
❌ Denied all net access.
❌ Denied all env access.
❌ Denied all env access.
❌ Denied all sys access.
❌ Denied all sys access.
❌ Denied all ffi access.
❌ Denied all ffi access.
```
Note: permission statuses are printed twice per permission use here. I didn't specifically enable this functionality, so I assume it's the default behaviour (wherever that code is).

File: runtime/permissions/prompter.rs
Comment (aapoalas) [inline runtime/permissions/prompter.rs:213]: Your rebase seems to have reverted the change that was done in https://github.com/denoland/deno/pull/17708 where the EMOJI was moved inside the `┌` blocks.

**Review by aapoalas — CHANGES_REQUESTED** [review]
> Rebased. Did you mean `Y`/`N` allow/deny all permissions? If so, `Y` and `N` behaviour now haven't been affected by this PR.

I meant both lower case and upper case Y, N, A or D working to select the corresponding option. ie. Indeed, no change for y/Y n/N has happened and that is good.

> Also, I tried the `a` behaviour, and it seems to work within expectations. Using `a` yields the following:

The unit tests indeed seem to work, but at least some real-life cases do not as shown by the example I gave. Try running the `tools/format.js` script in the deno repository and select `a` or `A` for "allow all".

File: runtime/permissions/prompter.rs
Comment (iuioiua) [reply runtime/permissions/prompter.rs:213]: 🥲

**Comment by iuioiua** [comment]
Ah, my bad, @aapoalas - I get you now. Yeah, I see that issue with `tools/format.js`... I'll see what's up with that.

**Comment by iuioiua** [comment]
I've got a few more things to do with this PR before it's ready. I'll aim to have it ready for 1.31.0.

File: runtime/permissions/prompter.rs
Comment (iuioiua) [reply runtime/permissions/prompter.rs]: Fixed!

**Comment by bartlomieju** [comment]
> I've got a few more things to do with this PR before it's ready. I aim to have it ready for 1.31.0.

@iuioiua do you need help with anything?

**Comment by iuioiua** [comment]
> @iuioiua do you need help with anything?

I think I have it figured out now, but I'll let you know if I do 🤙🏾

**Comment by iuioiua** [comment]
Previously, only `Deno.permissions.query()` calls were covered. Now, prompt interactions are covered. Preliminary tests on my local machine look good.

This is done by adding a `bool` to the return value of `PermissionState.check()` - `is_all`, which resembles whether the permission given in the prompt interaction is for one permission value (`y` or `n`) or all (`a` or `d`) of the same permission type (`env`, `read`, etc.)

Is this a sound approach?

**Comment by bartlomieju** [comment]
Tested locally, seems to be working very well.

> Previously, only `Deno.permissions.query()` calls were covered. Now, prompt interactions are covered. Preliminary tests on my local machine look good.
> 
> This is done by adding `is_all: bool` to the return value of `PermissionState.check()`, which resembles whether the permission given in the prompt interaction is for one permission value (`y` or `n`) or all (`a` or `d`) of the same permission type (`env`, `read`, etc.)
> 
> Is this a sound approach?

I'm not fully grepping it - `is_all` seems very ambigious, what are the situations where this flag would be `false`?

We should probably bike shed a bit on the prompt; I'm thinking that instead of `Allow? [y/n/a/d] (y = yes, allow; n = no, deny; a = yes to all, allow all; d = no to all, deny all) >` we could do `Allow? [y/n/a/d] (y = yes, allow; n = no, deny; a = allow all; d = deny all) >`

File: runtime/permissions/prompter.rs
Comment (lucacasonato) [inline runtime/permissions/prompter.rs:250]:
```suggestion
        'a' if is_unary => {
          // Passthrough if 'a' is used on the wrong permission
          ```

File: runtime/permissions/prompter.rs
Comment (lucacasonato) [inline runtime/permissions/prompter.rs:209]: Probably too long. Idea:

```
[y/n/a/d] (y = yes, allow; n = no, deny; a = always, allow all; d = never, deny all)
```

Comment (lucacasonato) [reply runtime/permissions/prompter.rs:209]: Also, what does `d` stand for? _deny_?

**Comment by lucacasonato** [comment]
I find the wording of the question a bit confusing. We need to make clear that _a = allow all **env vars**_ for example, not _all permissions in general_. Anyone have ideas to clarify?

**Comment by iuioiua** [comment]
Current:
```
[y/n/a/d] (y = yes, allow; n = no, deny; a = yes to all, allow all; d = no to all, deny all)
```

Suggestion (82 characters max.):
```
[y/n/a/d] (y = yes; n = no; a = yes to all write perms; d = no to all write perms)
```

File: runtime/permissions/prompter.rs
Comment (iuioiua) [reply runtime/permissions/prompter.rs:209]: Yep, `d` stands for deny all.

File: runtime/permissions/prompter.rs
Comment (iuioiua) [reply runtime/permissions/prompter.rs]: Change made.

**Comment by iuioiua** [comment]
> I'm not fully grepping it - `is_all` seems very ambigious, what are the situations where this flag would be `false`?

`is_all` could probably be renamed to `is_for_all_perm_values`. It'd be false when the user enters `y` for a single permission value. Does `is_for_all_perm_values` sound like a sufficient name?

**Comment by bartlomieju** [comment]
Hey @iuioiua we discussed this PR during today's CLI working group meeting. Since denying permissions in 99% will mean that program terminates, we want to remove the option to "deny all" and only leave "allow all".

Other suggestions:
- use "A" (capital "A") for "allow all"
- prompt:
```
Allow? [y/n/A] (y = yes, allow; n = no, deny; A = allow all read/write/env... permissions) >
```

**Comment by iuioiua** [comment]
> Hey @iuioiua we discussed this PR during today's CLI working group meeting. Since denying permissions in 99% will mean that program terminates, we want to remove the option to "deny all" and only leave "allow all".
> 
> Other suggestions:
> 
> * use "A" (capital "A") for "allow all"
> * prompt:
> 
> ```
> Allow? [y/n/A] (y = yes, allow; n = no, deny; A = allow all read/write/env... permissions) >
> ```

Sounds good. I'll do that.

**Comment by iuioiua** [comment]
Ok, done.

**Review by bartlomieju — APPROVED** [review]
Tested it locally and it works really nice, I think it's a huge ergonomics improvement for users. I will wait for some more reviews before merging.

LGTM, very nice work @iuioiua

**Pull request merged**
